File: config/manager_secret/rust/src/types/manifest.rs
//! What a component is allowed to ask for.

use std::collections::BTreeSet;

/// The logical secret names a component declares.
///
/// The provider refuses anything undeclared. Configuration gets least privilege from the build
/// graph -- a target that does not declare a section cannot see it -- but a secret cannot be a
/// build target, so the symmetric mechanism is this declaration, enforced at the provider
/// (design.md Decision 6).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Manifest {
    names: BTreeSet<String>,
}

impl Manifest {
    pub fn new<I, S>(names: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        Self { names: names.into_iter().map(Into::into).collect() }
    }

    pub fn declares(&self, name: &str) -> bool {
        self.names.contains(name)
    }

    /// Every declared name, sorted. Used to make a refusal actionable rather than merely a "no".
    pub fn declared(&self) -> Vec<&str> {
        self.names.iter().map(String::as_str).collect()
    }

    pub fn is_empty(&self) -> bool {
        self.names.is_empty()
    }
}

File: config/manager_secret/go/manifest.go
package secret

import "sort"

// Manifest is the set of logical secret names a component declares.
//
// The provider refuses anything undeclared. Configuration gets least privilege from the build
// graph -- a target that does not declare a section cannot see it -- but a secret cannot be a
// build target, so the symmetric mechanism is this declaration, enforced at the provider.
type Manifest struct {
	names map[string]struct{}
}

func NewManifest(names ...string) Manifest {
	set := make(map[string]struct{}, len(names))
	for _, n := range names {
		set[n] = struct{}{}
	}
	return Manifest{names: set}
}

func (m Manifest) Declares(name string) bool {
	_, ok := m.names[name]
	return ok
}

// Declared returns every declared name, sorted, so a refusal is stable and diffable rather than
// ordered by map iteration.
func (m Manifest) Declared() []string {
	out := make([]string, 0, len(m.names))
	for n := range m.names {
		out = append(out, n)
	}
	sort.Strings(out)
	return out
}

func (m Manifest) IsEmpty() bool { return len(m.names) == 0 }
